### PR TITLE
Initial microservice setup

### PR DESCRIPTION
**Pull request by AntoineDao**
This should sort out all the main endpoints and basic CRUD functionality for this service. Would rather not merge this in tonight/today as there are a couple more things I want to do here:

1. Properly document endpoints/schemas with examples
2. Replace bogus images in readmes
3. Write tests 🤖

### CODE REVIEW

File: app/api/endpoints.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+from app.models.security import User
+from app.core.jwt import get_user_info
+
+# TODO(): change input parameters to dependency
+from app.models.parameters import per_page_param, page_param
+from app.crud.database import (create_model_db, get_models_db, get_model_db, 
+    delete_model_db, get_faces_db, create_faces_db
+)
+
+HOST = 'https://api.pollination.cloud'
```
Comment (mostaphaRoudsari) [inline app/api/endpoints.py]: This should change to HOST_NAME and from `core/config`

File: app/api/endpoints.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+router = APIRouter()
+
+@router.get(
+    "/",
+    operation_id='sensor_grids_welcome',
```
Comment (mostaphaRoudsari) [inline app/api/endpoints.py]:
```suggestion
    operation_id='model_grids_welcome',
```

File: app/api/endpoints.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+)
+def get_models(
+    page: int = page_param,
+    per_page: int = per_page_param,
+    user: User = Depends(get_user_info)
+):
+    """Retrieve a list of sensor grids."""
+    models = get_models_db(page, per_page, user)
+    return [
+        ModelOut.parse_obj(model.to_model_out(HOST)) for model in models
```
Comment (mostaphaRoudsari) [inline app/api/endpoints.py]:
```suggestion
        ModelOut.parse_obj(model.to_model_out(HOST_NAME)) for model in models
```

File: app/api/endpoints.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    model: Model = Body(
+        ...,
+        description = "A Pollination model",
+        title = "Model",
+        example = {}), # TODO(): Add model example
+    user: User = Depends(get_user_info)
+):
+    """Create a new model."""
+    nid = create_model_db(model, user)
+    location = '%s/models/%s' % (HOST, nid)
```
Comment (mostaphaRoudsari) [inline app/api/endpoints.py]:
```suggestion
    location = '%s/models/%s' % (HOST_NAME, nid)
```

File: app/api/endpoints.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    return JSONResponse(
```
Comment (mostaphaRoudsari) [inline app/api/endpoints.py]:
```suggestion
    return JSONResponse(status_code=201,
```

File: app/api/endpoints.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    response_description='Retrieved',
+    response_model=ModelOut,
+    responses={**additional_error_responses}
+)
+def get_model(
+    id: UUID = Path(..., title = "Model id."),
+    user: User = Depends(get_user_info)
+):
+    """Retrieve a sensor grid."""
+    return ModelOut.parse_obj(get_model_db(id, user).to_model_out(HOST))
```
Comment (mostaphaRoudsari) [inline app/api/endpoints.py]:
```suggestion
    return ModelOut.parse_obj(get_model_db(id, user).to_model_out(HOST_NAME))
```

File: app/api/endpoints.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        ...,
+        description = "A list of Pollination model faces",
+        title = "Faces",
+        example = {}), # TODO(): Add model example
+    user: User = Depends(get_user_info)
+):
+    """Create new model faces."""
+    nids = create_faces_db(model, user)
+    location = '%s/models/%s/faces' % (HOST, nids)
+    return JSONResponse(
```
Comment (mostaphaRoudsari) [inline app/api/endpoints.py]: status_code=201,

File: app/crud/database.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    session = get_db_session()
+    session.add(new)
+    try:
+        session.commit()
+    except Exception as e:
+        session.rollback()
+        session.flush()
+        print(e)
+        # TODO(): Add logging
+        raise HTTPException(500, 'Failed to create the sensor-grid')
```
Comment (mostaphaRoudsari) [inline app/crud/database.py]:
```suggestion
        raise HTTPException(500, 'Failed to create the model')
```

File: app/crud/database.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    session.close()
+    if not model:
+        raise HTTPException(404, 'Model not found.')
+    return model
+
+
+def delete_model_db(id, user: User):
+    """Delete a model"""
+    session = get_db_session()
+    session.delete(get_model_db(id, user))
```
Comment (mostaphaRoudsari) [inline app/crud/database.py]: so this does also delete all the faces? very nice. 👌

File: app/models/database.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+import json
+
+Base = declarative_base()
+
+class Model(Base):
+    """Pollination model schema."""
+    __tablename__ = 'model'
+    id = sa.Column(sa.String, primary_key=True)
+    name = sa.Column(sa.String, nullable=False)
+    type = sa.Column(sa.String, nullable=False)
```
Comment (mostaphaRoudsari) [inline app/models/database.py]: Do we still need type?

File: app/models/database.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    convert_to_meters = sa.Column(sa.Numeric, default=1.0)
+    user_id = sa.Column(sa.String, nullable=False, default=False)
+    created_at = sa.Column(sa.DateTime, nullable=False,
+        default=datetime.datetime.utcnow())
+    face_count = sa.Column(sa.Integer, nullable=False)
+    faces = relationship("Face", back_populates='model', cascade='all, delete, delete-orphan')
```
Comment (mostaphaRoudsari) [inline app/models/database.py]: I see how cascade works now! very nice.

File: app/models/database.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+            face_count=len(data.faces)
+        )
+        faces_db = [
+            Face.from_dict(face, model.id, user)
+            for face in faces
+        ]
+        model.faces = faces_db
+        return model
+
+    def to_model_out(self, host):
```
Comment (mostaphaRoudsari) [inline app/models/database.py]: now you can get HOST_NAME directly from `core/config`

File: app/models/schemas.py [hunk 13]
```diff
@@ -215,4 +214,41 @@ class Model(BaseModel):
         default=1
     )
 
-    faces: List[Face]
+    faces: List[Face] = Schema(
```
Comment (mostaphaRoudsari) [inline app/models/schemas.py]: Shouldn't this support both Face and ShadeFace?

File: app/models/database.py (hunk 10 above)
Comment (AntoineDao) [reply app/models/database.py]: I'd keep it in to leave space for different kinds of model serialization? Especially if we're storing raw JSON no?

File: app/models/schemas.py (hunk 13 above)
Comment (AntoineDao) [reply app/models/schemas.py]: Yep!

File: app/models/database.py (hunk 10 above)
Comment (mostaphaRoudsari) [reply app/models/database.py]: It doesn't hurt. So you still think that we will have a different model type at some point?

File: app/core/config.py [hunk 14]
```diff
@@ -21,3 +21,4 @@
 
 PROJECT_NAME = os.getenv('PROJECT_NAME', 'Pollination Cloud')
 ALLOWED_HOSTS = CommaSeparatedStrings(os.getenv('ALLOWED_HOSTS', ''))
+HOST_NAME = os.getenv('HOST_NAME ', 'https://api.pollination.cloud')
```
Comment (mostaphaRoudsari) [inline app/core/config.py]: Hmmm. Interesting. I did set it to `localhost:8000` by default which is more useful for local testing and added it to .env file too.

File: app/models/database.py (hunk 10 above)
Comment (AntoineDao) [reply app/models/database.py]: Not 100% sure tbh...

File: app/core/config.py (hunk 14 above)
Comment (AntoineDao) [reply app/core/config.py]: Might be a good shout to switch that around actually. Ok if we do this later because I'm heading to be now? :smile:

**Comment by AntoineDao** [comment]
@mostaphaRoudsari I've added some tests now. Let me know what you think of the setup and teardown of the database and mocking of the server.

File: app/core/config.py (hunk 14 above)
Comment (mostaphaRoudsari) [reply app/core/config.py]: 👍

**Pull request merged**